### DISCUSSION TITLE
How to rewrite history using proxyWithHistory?

### DISCUSSION THREAD

**Discussion by pastelmind** [pmndrs/valtio#787] [Q&A]
How do I "rewrite" history in-place using `proxyWithHistory()`?

For example, given a history store that looks like this:

```ts
const store = proxyWithHistory({ count: 0 })
store.value.count = 5
store.value.count = 10
```

Current state will be:

```
index 0, { count: 0 }
index 1, { count: 5 }
index 2, { count: 10 }   <-- I am here
```

What do I need to update it to this?

```
index 0, { count: 0 }
index 1, { count: 5 }
index 2, { count: 11 }   <-- I am here
```

I tried this code:

```ts
store.undo()
store.value.count = 11
```

While this _seems_ to update `store.value`, it does not update the history snapshot at index 2; running an `undo()` followed by a `redo()` results in `store.value` being 10, not 11.

**Comment by pastelmind** [comment]
This seems to work, but I don't understand what it's doing:

```ts
store.undo()
store.value.count = 11
store.saveHistory()
```

**Comment by dai-shi** [comment]
Maybe, there's a bug in `proxyWithHistory`??

**Comment by lwhiteley** [comment]
Hi @pastelmind .

We have added `replace` and `remove` helper methods to `valtio-history`

You can try them out and give feedback/report issues if you can 

https://github.com/valtiojs/valtio-history/blob/main/packages/history-utility/docs/modules.md#returns

**Reply by pastelmind** [reply]
Hey, thanks!
Unfortunately I ended up writing own history implementation. In addition to resetting and replacing history, my history manager exposes `canUndo` and `canRedo` as getters rather than functions. It allows me to subscribe to these fields with `useSnapshot()`. I use this to selectively disable my undo/redo buttons.

I will look into `valtio-history` nevertheless. Again, thank you.

-----

By the way, is everything related to Valtio being moved under the "valtiojs" organization? This is news to me. Is there an official announcement, or a GitHub issue/discussion thread that can bring me up to speed?

**Reply by dai-shi** [reply]
https://twitter.com/dai_shi/status/1720266117124600252

I didn't open a discussion here. 😓 

By the way, do you notice that v2 is being ready?
